crates/typst/src/syntax/parser.rs

```diff
@@ -306,14 +312,38 @@ fn math_expr_prec(p: &mut Parser, min_prec: usize, stop: SyntaxKind) {
         p.wrap(m, SyntaxKind::Math);
     }
 
+    let mut primed = false;
+
     while !p.eof() && !p.at(stop) {
+        // if `stop` is hat, than prime is also stop
+        // a_b' works like a_b^'
+        if stop == SyntaxKind::Hat && p.at(SyntaxKind::Prime) {
+            break;
+        }
+
         if p.directly_at(SyntaxKind::Text) && p.current_text() == "!" {
             p.eat();
             p.wrap(m, SyntaxKind::Math);
             continue;
         }
 
+        if p.eat_if(SyntaxKind::Prime) {
+            // eat as many primes as possible
+            while p.eat_if(SyntaxKind::Prime) {}
+            primed = true;
+        }
+
+        // separate primes and superscripts to different attachments
+        if primed && p.current() == SyntaxKind::Hat {
```

Comment:
I don't really understand why `primed` is a thing across multiple iterations of this loop. for instance, `$a'_3'_4'$` behave somewhat weirdly (the first two primes merge, the third doesn't).

![empty](https://github.com/typst/typst/assets/17899797/3c733789-e35e-4954-868b-1952ed2bf5e0)

Reply:
I wasn't here for the original conversation and I understand that things were changed _towards_ this, but I don't really understand why the consensus was that parsing `a_b'` as `a'_b` instead of `a_(b')` is considered better. The latter doesn't lead to the issue we currently have, allows more things to be expressed without parenthesis and let's the source look more like the output.

One particular situation I've only just though of is `#let ab = $a_b$; $ab'$`, which would probably not be handled ideally because it's two attachments. But the same goes for `#let ab = $a_b$; $ab^2$` ...